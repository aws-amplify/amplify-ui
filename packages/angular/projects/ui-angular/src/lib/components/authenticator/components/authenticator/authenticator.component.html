<div data-amplify-authenticator [attr.data-variation]="variation">
  <div data-amplify-container>
    <amplify-slot name="header" [context]="context"></amplify-slot>
    <div
      data-amplify-router
      [attr.data-amplify-router-content]="hasTabs() ? undefined : ''"
    >
      <amplify-tabs
        (tabChange)="onTabChange()"
        *ngIf="route === 'signIn' || route === 'signUp'"
      >
        <amplify-tab-item
          [title]="signInTitle"
          [active]="route === 'signIn'"
          data-amplify-router-content
        >
          <!-- signIn component -->
          <amplify-slot
            name="sign-in"
            [context]="context"
            *ngIf="route === 'signIn'"
          >
            <amplify-sign-in></amplify-sign-in>
          </amplify-slot>
        </amplify-tab-item>
        <amplify-tab-item
          [title]="signUpTitle"
          [active]="route === 'signUp'"
          data-amplify-router-content
        >
          <!-- signUp component -->
          <amplify-slot
            name="sign-up"
            [context]="context"
            *ngIf="route === 'signUp'"
          >
            <amplify-sign-up></amplify-sign-up>
          </amplify-slot>
        </amplify-tab-item>
      </amplify-tabs>

      <!-- confirmSignUp content -->
      <amplify-slot
        name="confirm-sign-up"
        [context]="context"
        *ngIf="route === 'confirmSignUp'"
      >
        <amplify-confirm-sign-up></amplify-confirm-sign-up>
      </amplify-slot>

      <!-- confirmSignIn content -->
      <amplify-slot
        name="confirm-sign-in"
        [context]="context"
        *ngIf="route === 'confirmSignIn'"
      >
        <amplify-confirm-sign-in></amplify-confirm-sign-in>
      </amplify-slot>

      <!-- setupTotp content -->
      <amplify-slot
        name="setup-totp"
        [context]="context"
        *ngIf="route === 'setupTOTP'"
      >
        <amplify-setup-totp></amplify-setup-totp>
      </amplify-slot>

      <!-- forceNewPassword content -->
      <amplify-slot
        name="force-new-password"
        [context]="context"
        *ngIf="route === 'forceNewPassword'"
      >
        <amplify-force-new-password></amplify-force-new-password>
      </amplify-slot>

      <!-- resetPassword content -->
      <amplify-slot
        name="reset-password"
        [context]="context"
        *ngIf="route === 'resetPassword'"
      >
        <amplify-reset-password></amplify-reset-password>
      </amplify-slot>

      <!-- confirmResetPassword content -->
      <amplify-slot
        name="confirm-reset-password"
        [context]="context"
        *ngIf="route === 'confirmResetPassword'"
      >
        <amplify-confirm-reset-password></amplify-confirm-reset-password>
      </amplify-slot>

      <!-- verifyUser content -->
      <amplify-slot
        name="verify-user"
        [context]="context"
        *ngIf="route === 'verifyUser'"
      >
        <amplify-verify-user></amplify-verify-user>
      </amplify-slot>

      <!-- confirmVerifyUser content -->
      <amplify-slot
        name="confirm-verify-user"
        [context]="context"
        *ngIf="route === 'confirmVerifyUser'"
      >
        <amplify-confirm-verify-user></amplify-confirm-verify-user>
      </amplify-slot>
    </div>

    <amplify-slot name="footer" [context]="context"></amplify-slot>
  </div>
</div>

<!-- signedIn content is rendered outside authenticator so it's not styled by authenticator -->
<amplify-slot
  name="authenticated"
  [context]="context"
  *ngIf="route === 'authenticated'"
>
  <ng-content></ng-content>
</amplify-slot>
