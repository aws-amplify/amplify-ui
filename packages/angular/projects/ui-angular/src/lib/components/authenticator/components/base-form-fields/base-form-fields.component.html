<ng-container *ngFor="let formField of formFields">
  <amplify-form-field
    [name]="formField[0]"
    [formField]="formField[1]"
  ></amplify-form-field>
</ng-container>
