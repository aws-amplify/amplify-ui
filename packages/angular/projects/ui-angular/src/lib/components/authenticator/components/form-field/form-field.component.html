<div class="amplify-flex amplify-field" style="flex-direction: column">
  <!-- Country code field -->
  <amplify-phone-number-field
    *ngIf="isPhoneField()"
    [defaultCountryCode]="defaultCountryCode"
    [type]="inferType()"
    [name]="name"
    [label]="inferLabel()"
    [placeholder]="inferPlaceholder()"
    [required]="required"
    [initialValue]="initialValue"
    [disabled]="disabled"
    [labelHidden]="labelHidden"
    [autocomplete]="inferAutocomplete()"
  ></amplify-phone-number-field>

  <amplify-password-field
    *ngIf="isPasswordField()"
    [name]="name"
    [label]="inferLabel()"
    [placeholder]="inferPlaceholder()"
    [required]="required"
    [initialValue]="initialValue"
    [disabled]="disabled"
    [labelHidden]="labelHidden"
    [autocomplete]="inferAutocomplete()"
    (setBlur)="onBlur($event)"
  ></amplify-password-field>

  <amplify-text-field
    *ngIf="!isPasswordField() && !isPhoneField()"
    [type]="inferType()"
    [name]="name"
    [label]="inferLabel()"
    [placeholder]="inferPlaceholder()"
    [required]="required"
    [initialValue]="initialValue"
    [disabled]="disabled"
    [labelHidden]="labelHidden"
    [autocomplete]="inferAutocomplete()"
  ></amplify-text-field>

  <amplify-error *ngIf="error">
    {{ error }}
  </amplify-error>
</div>
