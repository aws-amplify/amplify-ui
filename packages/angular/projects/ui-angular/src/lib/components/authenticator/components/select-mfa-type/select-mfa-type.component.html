<form
    data-amplify-form
    data-amplify-authenticator-select-mfa-type
    (submit)="onSubmit($event)"
    (input)="onInput($event)"
>
    <fieldset
        [class]="classNames(ComponentClassName.Flex, 'amplify-authenticator__column')"
        data-amplify-fieldset
        [disabled]="authenticator.isPending"
    >
        <amplify-slot
            name="select-mfa-type-header"
            [context]="context"
        >
            <h3 [class]="classNames(ComponentClassName.Heading, 'amplify-heading--3')">{{ headerText }}</h3>
        </amplify-slot>
        <div
            *ngFor="let mfaType of allowedMfaTypes; index as i"
            [class]="classNames(ComponentClassName.Flex, ComponentClassName.Field, ComponentClassName.RadioGroupField)"
            role="radiogroup"
        >
            <label
                [class]="classNames(ComponentClassName.Flex, ComponentClassName.Radio)"
                [attr.data-amplify-radio-control-label]
            >
                <span [class]="classNames(ComponentClassName.Text, ComponentClassName.RadioLabel)">
                    {{getMfaTypeLabelByValue(mfaType)}}
                </span>
                <input
                    [class]="classNames(ComponentClassName.Input, ComponentClassName.FieldGroupControl, ComponentClassName.VisuallyHidden, ComponentClassName.RadioInput)"
                    [attr.data-amplify-radio-control-input]
                    [value]="mfaType"
                    [checked]="i === 0"
                    name="mfa_type"
                    type="radio"
                    required
                />
                <span
                    [class]="classNames(ComponentClassName.Flex, ComponentClassName.RadioButton)"
                    [attr.aria-hidden]="true"
                ></span>
            </label>
        </div>
        <button
            amplify-button
            variation="primary"
            fullWidth="true"
            type="submit"
            [isDisabled]="authenticator.isPending"
        >
            {{ confirmText }}
        </button>
        <button
            amplify-button
            size="small"
            variation="link"
            fontWeight="normal"
            fullWidth="true"
            (click)="authenticator.toSignIn()"
        >
            {{ backToSignInText }}
        </button>
        <amplify-error *ngIf="authenticator.error">
            {{ authenticator.error }}
        </amplify-error>
    </fieldset>
    <amplify-slot
        name="select-mfa-type-footer"
        [context]="context"
    ></amplify-slot>
</form>