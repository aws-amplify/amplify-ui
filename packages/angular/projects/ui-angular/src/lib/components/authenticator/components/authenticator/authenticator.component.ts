import {
  AfterContentInit,
  ChangeDetectorRef,
  Component,
  ContentChildren,
  Input,
  OnDestroy,
  OnInit,
  QueryList,
  TemplateRef,
  ViewEncapsulation,
} from '@angular/core';
import {
  AuthenticatorMachineOptions,
  authenticatorTextUtil,
  configureComponent,
  SocialProvider,
} from '@aws-amplify/ui';
import { AmplifySlotDirective } from '../../../../utilities/amplify-slot/amplify-slot.directive';
import { CustomComponentsService } from '../../../../services/custom-components.service';
import { AuthenticatorService } from '../../../../services/authenticator.service';
import { VERSION } from '../../../../../version';

const { getSignInTabText, getSignUpTabText } = authenticatorTextUtil;

@Component({
  selector: 'amplify-authenticator',
  templateUrl: './authenticator.component.html',
  providers: [CustomComponentsService], // make sure custom components are scoped to this authenticator only
  encapsulation: ViewEncapsulation.None,
})
export class AuthenticatorComponent
  implements OnInit, AfterContentInit, OnDestroy
{
  @Input() formFields: AuthenticatorMachineOptions['formFields'];
  @Input() initialState: AuthenticatorMachineOptions['initialState'];
  @Input() loginMechanisms: AuthenticatorMachineOptions['loginMechanisms'];
  @Input() services: AuthenticatorMachineOptions['services'];
  @Input() signUpAttributes: AuthenticatorMachineOptions['signUpAttributes'];
  @Input() socialProviders: SocialProvider[];
  @Input() variation: 'default' | 'modal';
  @Input() hideSignUp: boolean;

  @ContentChildren(AmplifySlotDirective)
  private customComponentQuery: QueryList<AmplifySlotDirective> = null;

  // translated texts
  public signInTitle = getSignInTabText();
  public signUpTitle = getSignUpTabText();

  private hasInitialized = false;
  private isHandlingHubEvent = false;
  private unsubscribeMachine: () => void;

  constructor(
    private authenticator: AuthenticatorService,
    private contextService: CustomComponentsService,
    private changeDetector: ChangeDetectorRef
  ) {}

  ngOnInit(): void {
    const {
      initialState,
      loginMechanisms,
      services,
      signUpAttributes,
      socialProviders,
      formFields,
    } = this;

    configureComponent({
      packageName: '@aws-amplify/ui-angular',
      version: VERSION,
    });

    const { initializeMachine } = this.authenticator;

    this.authenticator.hubSubject.subscribe(() => {
      /*
       * Hub events aren't properly caught by Angular, because they are
       * synchronous events. Angular tracks async network events and
       * html events, but not synchronous events like hub.
       *
       * On any notable hub events, we run change detection manually.
       */
      this.changeDetector.detectChanges();

      /*
       * Hub events that we handle can lead to multiple state changes:
       * e.g. `authenticated` -> `signOut` -> initialState.
       *
       * We want to ensure change detection runs all the way, until
       * we reach back to the initial state. Setting the below flag
       * to true to until we reach initial state.
       */
      this.isHandlingHubEvent = true;
    });

    /**
     * Subscribes to state machine changes and sends INIT event
     * once machine reaches 'setup' state.
     */
    this.unsubscribeMachine = this.authenticator.subscribe(() => {
      const { route } = this.authenticator;

      if (this.isHandlingHubEvent) {
        this.changeDetector.detectChanges();

        const initialStateWithDefault = initialState ?? 'signIn';

        // We can stop manual change detection if we're back to the initial state
        if (route === initialStateWithDefault) {
          this.isHandlingHubEvent = false;
        }
      }

      if (!this.hasInitialized && route === 'setup') {
        initializeMachine({
          initialState,
          loginMechanisms,
          services,
          signUpAttributes,
          socialProviders,
          formFields,
        });

        this.hasInitialized = true;
      }
    }).unsubscribe;

    /**
     * handling translations after content init, because authenticator and its
     * translations might be initialized before the main app's `ngOnInit` is run.
     */
    this.signInTitle = getSignInTabText();
    this.signUpTitle = getSignUpTabText();
  }

  /**
   * Lifecycle Methods
   */
  ngAfterContentInit(): void {
    this.contextService.customComponents = this.mapCustomComponents(
      this.customComponentQuery
    );
  }

  ngOnDestroy(): void {
    if (this.unsubscribeMachine) this.unsubscribeMachine();
  }

  /**
   * Class Functions
   */

  // context passed to "authenticated" slot
  public get context() {
    return this.authenticator.slotContext;
  }

  public get route() {
    return this.authenticator.route;
  }

  public onTabChange() {
    const route = this.authenticator.route;
    if (route === 'signIn') {
      this.authenticator.toSignUp();
    } else {
      this.authenticator.toSignIn();
    }
  }

  public hasTabs() {
    const { route } = this.authenticator;
    return route === 'signIn' || route === 'signUp';
  }

  public hasRouteComponent() {
    const { route } = this.authenticator;

    switch (route) {
      case 'authenticated':
      case 'idle':
      case 'setup':
      case 'signOut':
      case 'transition':
        return false;
      default:
        return true;
    }
  }

  private mapCustomComponents(
    componentQuery: QueryList<AmplifySlotDirective>
  ): Record<string, TemplateRef<any>> {
    if (!componentQuery) return {};
    const customComponents: Record<string, TemplateRef<any>> = {};
    componentQuery.forEach((component) => {
      customComponents[component.name] = component.template;
    });

    return customComponents;
  }
}
