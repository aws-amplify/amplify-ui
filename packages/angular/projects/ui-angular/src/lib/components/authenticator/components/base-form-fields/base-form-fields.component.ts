import { Component, HostBinding, Input, OnInit } from '@angular/core';
import {
  FormFieldComponents,
  FormFieldsArray,
  getSortedFormFields,
} from '@aws-amplify/ui';
import { AuthenticatorService } from '../../../../services/authenticator.service';

/**
 * Sorts the given formFields, then renders them in order.
 */
@Component({
  selector: 'amplify-base-form-fields',
  templateUrl: './base-form-fields.component.html',
})
export class BaseFormFieldsComponent implements OnInit {
  @Input() route: FormFieldComponents; // formFields to sort and render
  @HostBinding('style.display') display = 'contents';
  public formFields: FormFieldsArray = [];

  constructor(private authenticator: AuthenticatorService) {}

  ngOnInit(): void {
    const state = this.authenticator.authState;
    this.formFields = getSortedFormFields(this.route, state);
  }
}
