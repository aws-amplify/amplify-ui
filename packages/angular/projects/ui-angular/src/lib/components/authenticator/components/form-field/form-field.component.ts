import { Component, Input } from '@angular/core';
import {
  translate,
  countryDialCodes,
  FormFieldOptions,
  getErrors,
} from '@aws-amplify/ui';
import { nanoid } from 'nanoid';
import { AuthenticatorService } from '../../../../services/authenticator.service';

@Component({
  selector: 'amplify-form-field',
  templateUrl: './form-field.component.html',
})
export class FormFieldComponent {
  @Input() name: string; // name of the input field
  @Input() formField: FormFieldOptions; // form field options for this field

  public defaultCountryCodeValue: string;
  public countryDialCodesValue = countryDialCodes;
  public errorId = nanoid(12);

  constructor(private authenticator: AuthenticatorService) {}

  get ariaDescribedBy(): string | undefined {
    return this.hasError() ? this.errorId : undefined;
  }

  get errors(): string[] {
    const { validationErrors } = this.authenticator;
    return getErrors(validationErrors[this.name]);
  }

  public onBlur($event: Event): void {
    const { name } = <HTMLInputElement>$event.target;

    this.authenticator.updateBlur({ name });
  }

  isPasswordField(): boolean {
    return this.formField.type === 'password';
  }

  isPhoneField(): boolean {
    return this.formField.type === 'tel';
  }

  hasError(): boolean {
    return this.errors?.length > 0;
  }

  translate(phrase: string): string {
    return translate<string>(phrase);
  }
}
