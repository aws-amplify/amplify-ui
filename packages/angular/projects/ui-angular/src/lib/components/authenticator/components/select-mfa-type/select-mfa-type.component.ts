import { Component, OnInit } from '@angular/core';
import {
  getActorContext,
  getFormDataFromEvent,
  authenticatorTextUtil,
  ComponentClassName,
  classNames,
} from '@aws-amplify/ui';

import { AuthenticatorService } from '../../../../services/authenticator.service';

const {
  getConfirmText,
  getBackToSignInText,
  getSelectMfaTypeText,
  getSelectMfaTypeByChallengeName,
  getMfaTypeLabelByValue,
} = authenticatorTextUtil;

@Component({
  selector: 'amplify-select-mfa-type',
  templateUrl: './select-mfa-type.component.html',
})
export class SelectMfaTypeComponent implements OnInit {
  public headerText: string;
  public confirmText = getConfirmText();
  public backToSignInText = getBackToSignInText();
  public selectMfaTypeText = getSelectMfaTypeText();
  public getMfaTypeLabelByValue = getMfaTypeLabelByValue;

  public allowedMfaTypes = [];

  public classNames = classNames;
  public ComponentClassName = ComponentClassName;

  constructor(public authenticator: AuthenticatorService) {}

  public get context(): AuthenticatorService['slotContext'] {
    return this.authenticator.slotContext;
  }

  ngOnInit(): void {
    this.setHeaderText();
    this.setAllowedMfaTypes();
  }

  setHeaderText(): void {
    const actorContext = getActorContext(this.authenticator.authState);
    this.headerText = getSelectMfaTypeByChallengeName(
      actorContext.challengeName
    );
  }

  setAllowedMfaTypes(): void {
    const actorContext = getActorContext(this.authenticator.authState);
    this.allowedMfaTypes = actorContext.allowedMfaTypes;
  }

  onInput(event: Event): void {
    event.preventDefault();
    const { name, value } = event.target as HTMLInputElement;
    this.authenticator.updateForm({ name, value });
  }

  onSubmit(event: Event): void {
    event.preventDefault();
    this.authenticator.submitForm(getFormDataFromEvent(event));
  }
}
