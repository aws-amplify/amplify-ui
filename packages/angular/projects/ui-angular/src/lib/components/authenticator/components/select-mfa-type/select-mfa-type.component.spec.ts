import { AuthActorContext } from '@aws-amplify/ui';
import { fireEvent, render, screen } from '@testing-library/angular';

import { AuthenticatorService } from '../../../../services/authenticator.service';
import { SelectMfaTypeComponent } from './select-mfa-type.component';
import { AmplifySlotComponent } from '../../../../utilities/amplify-slot/amplify-slot.component';
import { BaseFormFieldsComponent } from '../base-form-fields/base-form-fields.component';
import { FormFieldComponent } from '../form-field/form-field.component';
import { ButtonComponent } from '../../../../primitives/button/button.component';
import { ErrorComponent } from '../../../../primitives/error/error.component';
import { TextFieldComponent } from '../../../../primitives/text-field/text-field.component';

const fieldName = 'mfa_type';
const emailRadioLabel = 'Email Message';
const totpRadioLabel = 'Authenticator App (TOTP)';
const emailFieldInput = { name: fieldName, value: 'EMAIL' };
const totpFieldInput = { name: fieldName, value: 'TOTP' };

const mockContext: Partial<AuthActorContext> = {
  challengeName: 'MFA_SETUP',
  allowedMfaTypes: ['EMAIL', 'TOTP'],
};

const mockAuthenticatorService = {
  isPending: false,
  authState: {
    context: {
      actorRef: { getSnapshot: () => ({ context: { ...mockContext } }) },
    },
  },
  updateForm: jest.fn(),
  submitForm: jest.fn(),
  toSignIn: jest.fn(),
  validationErrors: {},
  error: undefined,
};

const componentDeclarations = [
  ErrorComponent,
  SelectMfaTypeComponent,
  AmplifySlotComponent,
  BaseFormFieldsComponent,
  FormFieldComponent,
  ButtonComponent,
  TextFieldComponent,
];

jest.mock('nanoid', () => ({ nanoid: jest.fn(() => 'static') }));

describe('SelectMfaTypeComponent', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders as expected', async () => {
    const { container } = await render(SelectMfaTypeComponent, {
      declarations: componentDeclarations,
      providers: [
        {
          provide: AuthenticatorService,
          useValue: mockAuthenticatorService,
        },
      ],
    });

    expect(container).toMatchSnapshot();
  });

  it('sends change event on form input', async () => {
    await render(SelectMfaTypeComponent, {
      declarations: componentDeclarations,
      providers: [
        {
          provide: AuthenticatorService,
          useValue: mockAuthenticatorService,
        },
      ],
    });

    const totpRadioButton = await screen.findByText(totpRadioLabel);

    fireEvent.click(totpRadioButton);

    expect(mockAuthenticatorService.updateForm).toHaveBeenCalledWith(
      totpFieldInput
    );

    const emailRadioButton = await screen.findByText(emailRadioLabel);

    fireEvent.click(emailRadioButton);

    expect(mockAuthenticatorService.updateForm).toHaveBeenCalledWith(
      emailFieldInput
    );
  });

  it('sends submit event on form submit with default', async () => {
    await render(SelectMfaTypeComponent, {
      declarations: componentDeclarations,
      providers: [
        {
          provide: AuthenticatorService,
          useValue: mockAuthenticatorService,
        },
      ],
    });

    const submitButton = await screen.findByRole('button', { name: 'Confirm' });

    fireEvent.click(submitButton);

    expect(mockAuthenticatorService.submitForm).toHaveBeenCalledTimes(1);
    expect(mockAuthenticatorService.submitForm).toHaveBeenCalledWith({
      [emailFieldInput.name]: emailFieldInput.value,
    });
  });

  it('sends submit event on form submit with selection', async () => {
    await render(SelectMfaTypeComponent, {
      declarations: componentDeclarations,
      providers: [
        {
          provide: AuthenticatorService,
          useValue: mockAuthenticatorService,
        },
      ],
    });

    const totpRadioButton = await screen.findByText(totpRadioLabel);

    fireEvent.click(totpRadioButton);

    expect(mockAuthenticatorService.updateForm).toHaveBeenCalledWith(
      totpFieldInput
    );

    const submitButton = await screen.findByRole('button', { name: 'Confirm' });

    fireEvent.click(submitButton);

    expect(mockAuthenticatorService.submitForm).toHaveBeenCalledTimes(1);
    expect(mockAuthenticatorService.submitForm).toHaveBeenCalledWith({
      [totpFieldInput.name]: totpFieldInput.value,
    });
  });

  it('displays and error if present', async () => {
    await render(SelectMfaTypeComponent, {
      declarations: componentDeclarations,
      providers: [
        {
          provide: AuthenticatorService,
          useValue: { ...mockAuthenticatorService, error: 'mockError' },
        },
      ],
    });

    const mockError = await screen.findByText('mockError');

    expect(mockError).toBeInTheDocument();
  });

  it('handles back to sign in button as expected', async () => {
    await render(SelectMfaTypeComponent, {
      declarations: componentDeclarations,
      providers: [
        {
          provide: AuthenticatorService,
          useValue: mockAuthenticatorService,
        },
      ],
    });

    const backToSignInButton = await screen.findByRole('button', {
      name: 'Back to Sign In',
    });

    fireEvent.click(backToSignInButton);

    expect(mockAuthenticatorService.toSignIn).toHaveBeenCalledTimes(1);
  });

  it('disables the submit button if confirmation is pending', async () => {
    await render(SelectMfaTypeComponent, {
      declarations: componentDeclarations,
      providers: [
        {
          provide: AuthenticatorService,
          useValue: { ...mockAuthenticatorService, isPending: true },
        },
      ],
    });

    const submitButton = await screen.findByRole('button', { name: 'Confirm' });

    expect(submitButton).toBeDisabled();
  });
});
