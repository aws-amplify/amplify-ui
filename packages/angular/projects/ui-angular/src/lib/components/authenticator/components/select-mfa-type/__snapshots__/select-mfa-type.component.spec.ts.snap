// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectMfaTypeComponent renders as expected 1`] = `
<div
  id="root0"
>
  <form
    data-amplify-authenticator-select-mfa-type=""
    data-amplify-form=""
  >
    <fieldset
      class="amplify-authenticator__column amplify-flex"
      data-amplify-fieldset=""
    >
      <amplify-slot
        name="select-mfa-type-header"
        style="display: contents;"
      >
        <h3
          class="amplify-heading amplify-heading--3"
        >
          Multi-Factor Authentication Setup
        </h3>
        
        
      </amplify-slot>
      <div
        class="amplify-field amplify-flex amplify-radiogroupfield"
        role="radiogroup"
      >
        <label
          class="amplify-flex amplify-radio"
        >
          <span
            class="amplify-radio__label amplify-text"
          >
             Email Message 
          </span>
          <input
            class="amplify-field-group__control amplify-input amplify-radio__input amplify-visually-hidden"
            name="mfa_type"
            required=""
            type="radio"
            value="EMAIL"
          />
          <span
            aria-hidden="true"
            class="amplify-flex amplify-radio__button"
          />
        </label>
      </div>
      <div
        class="amplify-field amplify-flex amplify-radiogroupfield"
        role="radiogroup"
      >
        <label
          class="amplify-flex amplify-radio"
        >
          <span
            class="amplify-radio__label amplify-text"
          >
             Authenticator App (TOTP) 
          </span>
          <input
            class="amplify-field-group__control amplify-input amplify-radio__input amplify-visually-hidden"
            name="mfa_type"
            required=""
            type="radio"
            value="TOTP"
          />
          <span
            aria-hidden="true"
            class="amplify-flex amplify-radio__button"
          />
        </label>
      </div>
      
      <button
        amplify-button=""
        class="amplify-button amplify-button--fullwidth amplify-button--medium amplify-button--primary"
        data-fullwidth="true"
        data-size="medium"
        data-variation="primary"
        fullwidth="true"
        style="font-weight: normal;"
        type="submit"
        variation="primary"
      >
         Confirm 
      </button>
      <button
        amplify-button=""
        class="amplify-button amplify-button--fullwidth amplify-button--link amplify-button--small"
        data-fullwidth="true"
        data-size="small"
        data-variation="link"
        fontweight="normal"
        fullwidth="true"
        size="small"
        style="font-weight: normal;"
        type="button"
        variation="link"
      >
         Back to Sign In 
      </button>
      
    </fieldset>
    <amplify-slot
      name="select-mfa-type-footer"
      style="display: contents;"
    >
      
      
    </amplify-slot>
  </form>
</div>
`;
